lib/qmp.ml

```diff
@@ -143,10 +147,22 @@ let message_of_string str =
   in
 
   let event json =
-    let ts x = json |> U.member "timestamp" |> U.member x |> U.to_int in
-    { timestamp = (ts "seconds", ts "microseconds")
-    ; event     = json |> U.member "event" |> U.to_string
-    }
+    let event' json =
```

Comment:
this can stay as event(), and event_data() goes inside it to organise the scope of the subfunctions such as rtc_offset():
```
let event json =
+  let event_data event data =
+    let rtc_offset data = data |> U.member "offset" |> U.to_int in
+    match event with
+      | "RTC_CHANGE"  -> data |> rtc_offset |> fun x -> Some (RTC_CHANGE x)
+      | "POWERDOWN"  -> None
+      | "BLOCK_IO_ERROR"  -> None
+      | x -> Printf.fprintf stderr "unsupported event [%s]\n!" x; None
+  in

Reply:
Thanks, updated